### ISSUE TITLE
nghttp2_submit_data returns NGHTTP2_ERR_DATA_EXIST if HEADERS is submitted but not sent

### ISSUE THREAD

**Issue by lomnicky** [tatsuhiro-t/nghttp2#277]
If HEADERS frame is submitted for a promised stream but not sent yet, nghttp2_submit_data returns NGHTTP2_ERR_DATA_EXIST even though no DATA has been submitted yet. 

The code in nghttp2_session_add_item checks for emptiness of stream->item. As it contains a HEADERS frame, the code fails.

Induced with 1.0.4.

**Comment by tatsuhiro-t** [comment]
Assuming you are submitting response headers and body, consider to use nghttp2_submit_response(), which deals with this situation.  If you really need to call nghttp2_submit_headers() and nghttp2_submit_data() for some reason, call nghttp2_submit_data() inside nghttp2_on_frame_sent_callback for HEADERS.

**Comment by lomnicky** [comment]
I am creating a stream proxy and, hence, need to do it the low-level way. So any frame must be submitted only after the previous frame is sent? This is highly impractical. As there are queues already implemented by nghttp2, it means implementing them once again to support deferred submission.

**Comment by tatsuhiro-t** [comment]
You can do deferred submission of data chunk using nghttp2_submit_response() and NGHTTP2_ERR_DEFERRED (see https://nghttp2.org/documentation/types.html#c.nghttp2_data_source_read_callback).
In other words, instead of using nghttp2_submit_headers and nghttp2_submit_data to achieve deferred submission, use nghttp2_submit_response() and NGHTTP2_ERR_DEFERRED.
nghttpx uses this way to implement asynchronous server push.

**Comment by lomnicky** [comment]
So, please, update at least the documentation of nghttp2_submit_data.

**Comment by tatsuhiro-t** [comment]
Will do.  Currently, we have note section in https://nghttp2.org/documentation/nghttp2_submit_data.html, which only says about DATA frame.  I'll add HEADERS there too.

**Comment by lomnicky** [comment]
When I have other potential issues but would like to consult them before submitting, is it possible to talk to you somehow?

**Comment by tatsuhiro-t** [comment]
I have no problem to discuss anything in issue, whether or not it is a real issue.  So please submit the issue.

**Comment by tatsuhiro-t** [comment]
I added some text to clarify this issue: 693fba3b6473d18104dae32d0df970db5b6029d3

**Comment by tatsuhiro-t** [comment]
Closing since I think reported issue was resolved.
